docs/src/proposals/comprehensive-compute-fees.md

```diff
@@ -0,0 +1,97 @@
+---
+title: Comprehensive Compute Fees
+---
+
+## Motivation
+
+The current fee structure lacks a comprehensive account of the work required by
+a validator to process a transaction.  The fee structure is only based on the
+number of signatures in a transaction but is meant to account for the work that
+the validator must perform to validate each transaction.  The validator performs
+a lot more user-defined work than just signature verification.  The work to
+process a transaction typically includes signature verifications, account
+locking, account loading, and instruction processing.
+
+## Proposed Solution
+
+### New fee structure
+
+In addition to signature checking, the other work required to process an
+instruction should be taken into account.  To do this, the total fee should be
+an accumulated cost of all the work.  Each piece of work can be
+measured/accounted for in a way that makes sense for the type of work involved.
+
+A new fee structure could include:
+1. flat fee per signature
+2. flat fee for each write lock
+3. per-byte fee for the amount of data in each loaded account
+4. per-compute-unit fee for the number of compute units used to process an
+   instruction
+
+Fees 1-3 can be determined upfront before the message is processed.  #4 would
+need to be finally accounted for after the message is processed, but the payer's
+balance could be pre-deducted against a compute budget cap to ensure they have
+enough lamports to cover a max fee, and then any unused compute budget could be
+credited back after the message is fully processed.
+
+The goal of the fees is to cover the computation cost of processing a
+transaction.  Each of the above fee categories could be represented as a compute
+unit cost that, when added together, encompasses the entire cost of processing
+the transaction.  By calculating the total cost of the transaction the runtime
+can make better decisions on what transactions to process and when, as well as
+throw out transactions that exceed a cap.
+
+The per-compute unit fee doesn't have to be linear; developers could be
```

Comment:
I think we would want to like buckets with a fee rate for each one. The Transaction can request 100k compute units for fee `X`, then 200k for fee `2 * X`, etc. Then the bank-less leader can still rate the transactions accurately without execution.

Reply:
Yeah, for bank-less leader, we would want to know the caps upfront.  There could be a default cap as well as a requested cap (one of the instructions in the tx specifies it).  One challenge with the bank-less leader is that instructions using the default cap will probably be over-charged and I suspect relatively highly since they would be charged on a worst-case basis.  Also I was thinking that we might want to charge non-linearly so that higher cap buckets require relatively higher per-unit cost to encourage developers to optimize.  These kinds of questions probably require input from @ericlwilliams

Reply:
the cost model PR (https://github.com/solana-labs/solana/pull/16694) currently assigns fixed 'cost' artificially to instructions by type; implementing proposed fee structure there would be great

Reply:
Each bucket would have its own target and each compute unit range would be its own bucket, kinda like tax brackets

Reply:
> One challenge with the bank-less leader is that instructions using the default cap will probably be over-charged and I suspect relatively highly since they would be charged on a worst-case basis.

For the particular problem of page faults, which probably have the worst difference between average and worst case, I think allowing certain accounts to pay a steeper rent to be "Ram-Resident" could make sense, and accessing those would require smaller fees. Practically, the mechanism could be to never flush these accounts from the read only cache.